### PR TITLE
Tweaks to the file format spec for the compact index

### PR DESCRIPTION
**Pull request by dcoutts**
* Reorder components to allow incremental writes, with CRC calculations
* Add the LTP bit array
* Put sizes at the end to simplify reading

### CODE REVIEW

File: doc/format-run.md [hunk 1]
```diff
@@ -185,24 +185,28 @@ whole key per page: we can typically store just 4 bytes (32bits) per page. This
...
-prior to the completion of the index. The range finder array cannot be written
-until index completion, but its size is known in advance and so space in the
-file can be reserved. The clash indicator bit vector has the same number of
-elements as the primary array, but is of course 32 times smaller, so it is
-reasonable to keep this in memory while the index is constructed and flush it
-upon completion. Similarly, the clash map is small and it can be written upon
-index completion.
+written out to disk incrementally. All the remaining parts can only be written
+upon the completion of the index. These parts must be kept in memory while the
+index is constructed and can be flush upon completion.
```
Comment (mheinzel) [inline doc/format-run.md:203]:
```suggestion
index is constructed and can be flushed upon completion.
```

Comment (jorisdral) [inline doc/format-run.md:203]:
```suggestion
index is constructed and can be flushed upon completion.
```

File: doc/format-run.md [hunk 2]
```diff
@@ -185,24 +185,28 @@ whole key per page: we can typically store just 4 bytes (32bits) per page. This
...
 2. a range finder array, of 2^n+1 entries of 32bit each (n = range finder bits)
-3. a primary array of 32bit words, one entry per page in the index
-4. a clash indicator bit vector, one bit per page in the index
+3. a clash indicator bit vector, one bit per page in the index
+4. a larger-than-page indicator bit vector, one bit per page in the index
 5. a clash map, mapping each page with a clash indicator to the full minimum
    key for the page
+6. the number of range finder bits (0..16) (as a 8bit value)
+7. the number of pages in the primary array (as a 64bit value)
+8. the number of keys in the corresponding key/ops file (as a 64bit value)
```
Comment (jorisdral) [inline doc/format-run.md:196]: What would we use this number for? I think we only need to know the number of pages

**Review by jorisdral — APPROVED** [review]

**Review by mheinzel — APPROVED** [review]
One think possibly worth adding: What is the padding/alignment of the components? Byte-aligned?

File: doc/format-run.md (hunk 2 above)
Comment (dcoutts) [reply doc/format-run.md:196]: It's not strictly necessary for the index format itself, but we were discussing the other day that we wanted to know this number, and we needed somewhere convenient to put it.

**Comment by dcoutts** [comment]
> One thing possibly worth adding: What is the padding/alignment of the components? Byte-aligned?

Done. I've tweaked it so that if we really want to we have the option to mmap the whole file and access all but the clash map directly with natural alignment.

**Review by jorisdral — APPROVED** [review]

**Review by mheinzel — APPROVED** [review]

**Pull request merged**